L74-java-channel-creds.md

```diff
@@ -0,0 +1,489 @@
+      CompositeChannelCredentials compCreds = (CompositeChannelCredentials) creds;
+      return from(compCreds.getChannelCredentials())
+          .withCallCredentials(compCreds.getCallCredentials());
+
+    } else if (creds instanceof NettyChannelCredentials) {
+      NettyChannelCredentials nettyCreds = (NettyChannelCredentials) creds;
+      return Result.negotiator(nettyCreds.getNegotiator());
+
+    } else if (creds instanceof ChoiceChannelCredentials) {
+      ChoiceChannelCredentials choiceCreds = (ChoiceChannelCredentials) creds;
+      StringBuilder error = new StringBuilder();
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        Result result = from(innerCreds);
+        if (!result.isError()) {
+          return result;
+        }
+        error.append("; ");
+        error.append(result.getError());
+      }
+      return Result.error(error.substring(2));
+
+    } else {
+      return Result.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  public final class Result {
+    public static Result error(String error) {...}
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted. There is also currently no known
+use-case for `ChoiceServerCredentials`, so it will be omitted.
+
+```java
+// instead of:
+server = ServerBuilder.forPort(443)
+    .useTransportSecurity(new File("cert.pem"), new File("cert.key"))
+    .build().start();
+server = ServerBuilder.forPort(8080)
+    .build().start();
+// the user would now:
+ChannelCredentials tlsCreds = new TlsServerCredentials(
+    new File("cert.pem"), new File("cert.key"));
+server = Grpc.newServerBuilderForPort(443, tlsCreds)
+    .build().start();
+server = Grpc.newServerBuilderForPort(8080, new InsecureChannelCredentials())
+    .build().start();
+```
+
+
+```java
+package io.grpc;
+
+public abstract class ServerCredentials {}
+```
+
+Since servers using TLS require a certificate, the `TlsServerCredentials`
+cannot have the simple no-arg configuration of `TlsChannelCredentials`. It will
+provide convenience constructors for providing a certificate and unencrypted
+key. If any other configuration is necessary, the `Builder` would be used.
+
+```java
+package io.grpc;
+
+public final class TlsServerCredentials extends ServerCredentials {
+  /**
+   * Creates an instance using provided certificate chain and private key.
+   * Generally they should be PEM-encoded and the key is an unencrypted PKCS#8
+   * key (file headers have "BEGIN CERTIFICATE" and "BEGIN PRIVATE KEY").
+   */
+  public TlsServerCredentials(File certChain, File privateKey)
```

Comment:
nit: `keyPassword` param needed for this and the next one?

Reply:
No, this is mirroring the current `serverBuilder.transportSecurity(...)` methods which don't support the password and are pretty common. If you need to provide a Key, use the Builder. I don't find specifying the password to be more rare, since it doesn't generally provide much protection because the password is generally plaintext and no more difficult to acquire than the private key file.